src/xray/askui-xray-step-reporter.ts

```diff
@@ -0,0 +1,155 @@
+import { Reporter, Step, StepStatus, ReporterConfig } from 'askui';
+import { convertPngDataUrlToBase64 } from '../utils/image-reporting-utils';
+import path from 'path';
+import fs from 'fs';
+
+/*
+ * Possible XRay status: PASS, FAIL, TODO
+ */
+enum Status {
+  PASS = 'PASS',
+  FAIL = 'FAIL',
+  TODO = 'TODO',
+}
+
+type XRayTestObject = {
+  testKey: string;
+  status?: Status;
+  start?: string;
+  finish?: string;
+  steps?: XRayStep[];
+}
+
+type XRayEvidence = {
+  data: string;
+  filename: string;
+  contentType: string;
+}
+
+type XRayStep = {
+  status?: Status;
+  actualResult?: string;
+  evidences: XRayEvidence[];
+}
+
+function mapAskuiToXrayStepStatus(status: StepStatus): Status {
+  switch (status) {
+    case 'passed':
+      return Status.PASS;
+    case 'failed':
+      return Status.FAIL;
+    case 'erroneous':
+      return Status.FAIL;
+    default:
+      return Status.TODO;
+  }
+}
+
+type StatusJest = 'success' | 'failure';
+
+function mapJestToXrayStatus(status: StatusJest): Status {
+  switch (status) {
+    case 'success':
+      return Status.PASS;
+    default:
+      return Status.FAIL;
+  }
+}
+
+export class AskUIXRayStepReporter implements Reporter {
+  config?: ReporterConfig;
+  result: XRayTestObject[] = [];
+  outputDirectory: string;
+
+  constructor(
+    config?: ReporterConfig,
+    outputDirectory = 'xray-report'
+    ) {
+    if (config !== undefined) {
+      this.config = config;
+    }
+    this.outputDirectory = outputDirectory;
+  }
+
+  async createNewTestEntry(
+    testTitle: string
+  ): Promise<void> {
+    this.result.push(
+      {
+        testKey: testTitle,
+        start: new Date().toISOString(),
+        steps: []
+      }
+    );
+  }
+
+  async finishTestEntry(
+    testStatus: StatusJest
+  ): Promise<void> {
+    if (this.result.length > 0) {
+      const testEntry = this.result.pop();
+      if (testEntry !== undefined) {
+        testEntry.finish = new Date().toISOString();
+        testEntry.status = mapJestToXrayStatus(testStatus);
+        this.result.push(testEntry);
+      }
+    }
```

Comment:
Isn't the length checking and result popping and pushing unnecessary? Shouldn't it be an exception if the testEntry is undefined? Same for the other functions below with a similar logic.

Reply:
I can omit the `length` check, but not the check for `undefined` because `pop()` return type is `XrayTestObject | undefined`.

Not sure what to do here tbh.